external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/CachedKafkaConsumer.scala

```diff
@@ -47,40 +51,229 @@ private[kafka010] case class CachedKafkaConsumer private(
+    } else {
+      // ------------------------------------------------------------------------------
+      //      ^           ^                       ^                                 ^
+      //      |           |                       |                                 |
+      //   offset   earliestOffset   min(untilOffset,latestOffset)   max(untilOffset, latestOffset)
+      val warningMessage =
+        s"""
+           |The current available offset range is [$earliestOffset, $latestOffset).
+           | Offset ${offset} is out of range, and records in [$offset, $earliestOffset) will be
+           | skipped (GroupId: $groupId, TopicPartition: $topicPartition).
+           | $INSTRUCTION_FOR_FAIL_ON_DATA_LOSS_FALSE
+        """.stripMargin
+      logWarning(warningMessage)
+      get(earliestOffset, untilOffset, pollTimeoutMs, failOnDataLoss = false)
+    }
+  }
 
-    if (record.offset != offset) {
-      logInfo(s"Buffer miss for $groupId $topicPartition $offset")
-      seek(offset)
-      poll(pollTimeoutMs)
-      assert(fetchedData.hasNext(),
-        s"Failed to get records for $groupId $topicPartition $offset " +
-          s"after polling for $pollTimeoutMs")
-      record = fetchedData.next()
-      assert(record.offset == offset,
-        s"Got wrong record for $groupId $topicPartition even after seeking to offset $offset")
+  /**
+   * Get the earliest record in [offset, untilOffset) from the fetched data. If there is no such
+   * record, return null. Note that this must be called after some data has already been fetched
+   * using poll.
+   */
+  private def getRecordFromFetchedData(
+      offset: Long,
+      untilOffset: Long,
+      failOnDataLoss: Boolean): ConsumerRecord[Array[Byte], Array[Byte]] = {
+    if (!fetchedData.hasNext()) {
+      // We cannot fetch anything after `poll`. Two possible cases:
+      // - `earliestOffset` is `offset` but there is nothing for `earliestOffset` right now.
+      // - Cannot fetch any data before timeout.
+      // Because there is no way to distinguish, just skip the rest offsets in the current
+      // partition.
+      val message =
+        if (failOnDataLoss) {
+          s"""
+             |Cannot fetch record for offset ${offset}
+             | (GroupId: $groupId, TopicPartition: $topicPartition).
+             | $INSTRUCTION_FOR_FAIL_ON_DATA_LOSS_TRUE""".stripMargin
+        } else {
+          s"""
+             |Cannot fetch record for offset ${offset}. Records in [$offset, $untilOffset) will be
+             | skipped (GroupId: $groupId, TopicPartition: $topicPartition).
+             | $INSTRUCTION_FOR_FAIL_ON_DATA_LOSS_FALSE""".stripMargin
+        }
+      reportDataLoss(failOnDataLoss, message)
+      reset()
+      null
+    } else {
+      val record = fetchedData.next()
+      if (record.offset >= untilOffset) {
+        // This may happen when records are aged out.
+        val message =
+          if (failOnDataLoss) {
```

Comment:
There is ALREADY a if statement in reportDataLoss on failOnDataLoss. having another one, looks too complicated. i think best to not have an if statement here. and same for other places. this too much complication for documentation.